### PR TITLE
Prevent usage of astro:content in the client

### PR DESCRIPTION
**Pull request by matthewp**
## Changes

- Previously we warned users who imported it in the client, this changes that to a runtime error.

## Testing

- Hard to test given the build breaks, but willing to do so if we think it's important.

## Docs

- https://github.com/withastro/docs/commit/c9828f31781747099b307ba09d3c20f0cea6389d

### CODE REVIEW

File: packages/astro/src/content/vite-plugin-content-virtual-mod.ts [hunk 1]
```diff
@@ -360,7 +372,7 @@ export async function generateLookupMap({
 					const contentEntryType = contentEntryConfigByExt.get(extname(filePath));
 					if (!contentEntryType) throw UnexpectedLookupMapError;
 
-					const { id, slug: generatedSlug } = await getContentEntryIdAndSlug({
+					const { id, slug: generatedSlug } = getContentEntryIdAndSlug({
```
Comment (matthewp) [inline packages/astro/src/content/vite-plugin-content-virtual-mod.ts:375]: this is an unrelated change, but this `await` is on a sync function, so I removed.

File: packages/astro/src/content/vite-plugin-content-virtual-mod.ts [hunk 2]
```diff
@@ -253,18 +259,24 @@ export async function generateContentEntryFile({
...
-			.replace("'@@RENDER_ENTRY_GLOB_PATH@@'", renderEntryGlobResult)
-			.replace('/* @@LOOKUP_MAP_ASSIGNMENT@@ */', `lookupMap = ${JSON.stringify(lookupMap)};`) +
-		(isClient
-			? `
-console.warn('astro:content is only supported running server-side. Using it in the browser will lead to bloated bundles and slow down page load. In the future it will not be supported.');`
-			: '');
+	let virtualModContents: string;
+	if(isClient) {
+		// Empty exports is probably enough to break the build.
+		// If for some reason its being imported as a side-effect ala `import 'astro:content';` then throw an error message.
```
Comment (Fryuni) [inline packages/astro/src/content/vite-plugin-content-virtual-mod.ts:265]: If is is imported dynamically like `await import('astro:content');` it won't fail the build either.

File: .changeset/neat-dots-hear.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
+---
+'astro': major
+---
+
+Prevent usage of astro:content in the client
```
Comment (ematipico) [inline .changeset/neat-dots-hear.md:5]:
```suggestion
Prevent usage of `astro:content` in the client
```

File: packages/astro/src/content/vite-plugin-content-virtual-mod.ts [hunk 4]
```diff
@@ -253,18 +259,24 @@ export async function generateContentEntryFile({
...
+		virtualModContents = `export {};
+throw new Error('astro:content is only supported running server-side.');`;
+
+		logger.error('content', 'astro:content is only supported running server-side.');
```
[the remark this answers is not in the document]
Comment (florian-lefebvre) [reply packages/astro/src/content/vite-plugin-content-virtual-mod.ts:269]: There's a `ServerOnlyModule` error btw

File: packages/astro/src/content/vite-plugin-content-virtual-mod.ts [hunk 5]
```diff
@@ -253,18 +253,27 @@ export async function generateContentEntryFile({
...
-			.replace('/* @@LOOKUP_MAP_ASSIGNMENT@@ */', `lookupMap = ${JSON.stringify(lookupMap)};`) +
-		(isClient
-			? `
-console.warn('astro:content is only supported running server-side. Using it in the browser will lead to bloated bundles and slow down page load. In the future it will not be supported.');`
-			: '');
+	let virtualModContents: string;
+	if(isClient) {
+		// Empty exports is probably enough to break the build.
+		// If for some reason its being imported as a side-effect ala `import 'astro:content';` then throw an error message.
+		virtualModContents = `export {};
```
Comment (florian-lefebvre) [inline packages/astro/src/content/vite-plugin-content-virtual-mod.ts:260]: Is this really needed? I think for `astro:env` I just throw without returning any content

Comment (matthewp) [reply packages/astro/src/content/vite-plugin-content-virtual-mod.ts]: good point, removed

**Review by florian-lefebvre — APPROVED** [review]

File: packages/astro/src/content/vite-plugin-content-virtual-mod.ts (hunk 1 above)
Comment (louismanhnguyen1998) [reply packages/astro/src/content/vite-plugin-content-virtual-mod.ts:367]: string

File: packages/astro/src/content/vite-plugin-content-virtual-mod.ts (hunk 2 above)
Comment (louismanhnguyen1998) [reply packages/astro/src/content/vite-plugin-content-virtual-mod.ts]: import

**Pull request merged**